### PR TITLE
Better error handling

### PR DESCRIPTION
**Pull request by Etenil**
Added error handling on failed execution with a custom exception that
bears most of the info regarding the failed query.

Also added a new FluentUtils function that takes a query and a bunch
of parameters and outputs them roughly the same way they'd be queried,
that's very useful for debugging.

### CODE REVIEW

File: FluentPDO/BaseQuery.php [hunk 1]
```diff
@@ -114,9 +117,15 @@ public function execute() {
 		}
 
 		$time = microtime(true);
-		if ($result && $result->execute($parameters)) {
+		if ($result && @$result->execute($parameters)) {
 			$this->time = microtime(true) - $time;
-		} else {
+		} else { // Oops, something went wrong with the query.
+            $info = $result->errorInfo();
+            $this->last_error = new FluentQueryException(
```
Comment (hoshsadiq) [inline FluentPDO/BaseQuery.php]: Exceptions are meant to be thrown, not stored.

Comment (Etenil) [inline FluentPDO/BaseQuery.php]: Fair enough. However currently they're neither thrown nor stored.

If I change this code to throw the exception instead, would that be satisfactory?

Comment (hoshsadiq) [inline FluentPDO/BaseQuery.php]: I just noticed the code, it's @lichtner that would have to accept it. If you want me to, I can review the rest of the code.

Comment (motin) [inline FluentPDO/BaseQuery.php]: @Etenil Either throw the exception or refactor FluentQueryException to FluentQueryError or similar

**Comment by motin** [comment]
Looking forward to better error handling! Silent errors are just the worst...

File: FluentPDO/BaseQuery.php [hunk 2]
```diff
@@ -114,9 +117,15 @@ public function execute() {
 		}
 
 		$time = microtime(true);
-		if ($result && $result->execute($parameters)) {
+		if ($result && @$result->execute($parameters)) {
```
Comment (motin) [inline FluentPDO/BaseQuery.php]: Please don't suppress errors with @. There is almost never any valid reason for this

**Comment by motin** [comment]
I think this implementation is currently a bit to biased. What about implementing an error handler similar to how the debug option is enabled? I created a PR here: https://github.com/lichtner/fluentpdo/pull/110

**Comment by victor-torres** [comment]
"Exceptions are meant to be thrown" +1

**Comment by cbornhoft** [comment]
`FluentPDO Exception: Query: Select * from table where this = :that; Parameters:[that => 'those']`

File: FluentPDO/FluentUtils.php [hunk 3]
```diff
@@ -23,4 +23,23 @@ public static function formatQuery($query) {
...
\ No newline at end of file
+
+    /** Fills in query placeholders with their parameter values; useful only for debugging.
+     * @param string $query
+     * @param array parameters
+     */
+    public static function populate($query, array $parameters) {
+        foreach ($parameters as $parameter_value) {
+            // Quoting mechanism.
+            if(!in_array($parameter_value, array(
```
Comment (hoshsadiq) [inline FluentPDO/FluentUtils.php]: You're redefining the array on every iteration. Bad idea, move the array outside the loop.

File: FluentPDO/FluentUtils.php [hunk 4]
```diff
@@ -23,4 +23,23 @@ public static function formatQuery($query) {
...
 	}
-}
\ No newline at end of file
+
+    /** Fills in query placeholders with their parameter values; useful only for debugging.
+     * @param string $query
+     * @param array parameters
+     */
+    public static function populate($query, array $parameters) {
+        foreach ($parameters as $parameter_value) {
```
Comment (hoshsadiq) [inline FluentPDO/FluentUtils.php]: Use camelCasing

File: FluentPDO/BaseQuery.php [hunk 5]
```diff
@@ -116,7 +116,13 @@ public function execute() {
 		$time = microtime(true);
 		if ($result && $result->execute($parameters)) {
 			$this->time = microtime(true) - $time;
-		} else {
+		} else { // Oops, something went wrong with the query.
+            $info = $result->errorInfo();
```
Comment (hoshsadiq) [inline FluentPDO/BaseQuery.php]: Check indentation

File: FluentPDO/BaseQuery.php [hunk 6]
```diff
@@ -198,6 +204,17 @@ public function getQuery($formated = true) {
 		return $query;
 	}
 
+    /** Get the full query with parameters (naively) replaced
```
Comment (hoshsadiq) [inline FluentPDO/BaseQuery.php]: Check indentation

File: FluentPDO/FluentQueryException.php [hunk 7]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+/** Base query exception; is triggered by an incorrect query.
+ */
+class FluentQueryException extends Exception
+{
+    private $query;
+
+    public function __construct($query, $message, $code) {
```
Comment (hoshsadiq) [inline FluentPDO/FluentQueryException.php]: phpdoc pls

File: FluentPDO/FluentQueryException.php [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    public function __tostring() {
+        return sprintf(
+            "Fluent query failed with error %s and message '%s' in %s:%d\nQuery: %s",
+            $this->getCode(), $this->getMessage(), $this->getFile(), $this->getLine(), $this->getQuery()
+        );
+    }
+
+    public function getQuery() {
```
Comment (hoshsadiq) [inline FluentPDO/FluentQueryException.php]: missing  phpdoc

File: FluentPDO/FluentQueryException.php [hunk 9]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+/** Base query exception; is triggered by an incorrect query.
+ */
+class FluentQueryException extends Exception
+{
+    private $query;
```
Comment (hoshsadiq) [inline FluentPDO/FluentQueryException.php]: missing  phpdoc

**Comment by odan** [comment]
You just have to set the correct parameters and PDO will throw a PDOException and set its properties to reflect the error code and error information.

`$pdo = new PDO('mysql:host=127.0.0.1;dbname=test;charset=utf8', 'root', '', [PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION]);`
`$fpdo = new FluentPDO($pdo);`

Details:
http://php.net/manual/en/pdo.error-handling.php

**Pull request closed without merging**